src/core/lib/support/any.h

```diff
@@ -0,0 +1,215 @@
+    new (storage) TP(New<T>(value));
+  }
+
+  static void MoveInit(T&& value, void* storage) {
+    new (storage) TP(New<T>(std::move(value)));
+  }
+
+  static T* Get(void* storage) { return static_cast<TP*>(storage)->get(); }
+
+  static const T* Get(const void* storage) {
+    return static_cast<const TP*>(storage)->get();
+  }
+
+  static void Copy(const void* from, void* to) {
+    Init(**static_cast<const TP*>(from), to);
+  }
+
+  static void Move(void* from, void* to) {
+    MoveInit(std::move(**static_cast<TP*>(from)), to);
+  }
+
+  static void CopyT(const void* from, void* to) {
+    **static_cast<TP*>(to) = **static_cast<const TP*>(from);
+  }
+
+  static void MoveT(void* from, void* to) {
+    *static_cast<TP*>(to) = std::move(*static_cast<TP*>(from));
+  }
+
+  static void Destroy(void* value) { static_cast<TP*>(value)->~TP(); }
+};
+
+template <class T>
+const VTable TypeInfo<T, true>::vtable = {
+    Copy, Move, CopyT, MoveT, Destroy,
+};
+
+template <class T = void>
+struct Null {
+  static const VTable vtable;
+
+  static void Copy(const void* from, void* to) {}
+  static void Move(void* from, void* to) {}
+  static void Destroy(void* value) {}
+};
+
+template <class T>
+const VTable Null<T>::vtable = {
+    Copy, Move, Copy, Move, Destroy,
+};
+
+}  // namespace any_impl
+
+template <size_t kInlineSize = sizeof(void*)>
+class Any {
+  template <class T>
+  using TI = any_impl::TypeInfo<T, (sizeof(T) > kInlineSize)>;
+
+ public:
+  Any() : vtable_(&any_impl::Null<>::vtable) {}
+  ~Any() {
+    static_assert(kInlineSize >= sizeof(void*),
+                  "Inlined data must be at least sizeof(void*)");
+    vtable_->destroy(&storage_);
+  }
+
+  Any(const Any& other) : vtable_(other.vtable_) {
+    vtable_->copy_init(&other.storage_, &storage_);
+  }
+
+  Any(Any& other) : vtable_(other.vtable_) {
+    vtable_->copy_init(&other.storage_, &storage_);
+  }
+
+  Any(Any&& other) : vtable_(other.vtable_) {
+    vtable_->move_init(&other.storage_, &storage_);
+  }
+
+  template <class T>
+  Any(const T& value) : vtable_(&TI<T>::vtable) {
+    TI<T>::Init(value, &storage_);
+  }
+
+  template <class T>
+  Any(T&& value) : vtable_(&TI<T>::vtable) {
+    TI<T>::MoveInit(std::move(value), &storage_);
+  }
+
+  Any& operator=(const Any& other) {
+    if (other.vtable_ == vtable_) {
+      vtable_->copy(&other.storage_, &storage_);
+    } else {
+      vtable_->destroy(&storage_);
+      vtable_ = other.vtable_;
+      vtable_->copy_init(&other.storage_, &storage_);
+    }
+    return *this;
+  }
+
+  Any& operator=(Any&& other) {
+    if (other.vtable_ == vtable_) {
+      vtable_->move(&other.storage_, &storage_);
+    } else {
+      vtable_->destroy(&storage_);
+      vtable_ = other.vtable_;
+      vtable_->move_init(&other.storage_, &storage_);
+    }
+    return *this;
+  }
+
+  template <class T>
+  T* as() {
+    return vtable_ == &TI<T>::vtable ? TI<T>::Get(&storage_) : nullptr;
```

Comment:
Do you plan on using this nullptr value to mean something? Otherwise, would rather an 1abort()1 if someone asks to use the wrong type